### PR TITLE
Fix prefetching the same GQL type for batch request

### PR DESCRIPTION
**Pull request by grynchuk**
I have issue with getting wrong prefetch data while query in batch mode.  In short, first query prefetch specific data and set it to prefetch buffer, second query use previous query prefetch buffer instead of using its own and get incomplete data set. I have to note that this issue is relates to prefetching of nested data, see test for mo details.

Request: 
```
[
  {
    "query": "query { contact (name: "Joe") { name  posts { id  title } } }",
  },
  {
    "query": "query { company (id: "1"){ name contact  { name  posts { id  title } } } }",
  }
]
```

Response
Actual:
```
[{"data":{"contact":{"name":"Joe","posts":[{"id":1,"title":"First Joe post"}]}}},{"data":{"company":{"name":"Company","contact":{"name":"Kate","posts":null}}}}]
```
Expected:
```
[{"data":{"contact":{"name":"Joe","posts":[{"id":1,"title":"First Joe post"}]}}},{"data":{"company":{"name":"Company","contact":{"name":"Kate","posts":[{"id":3,"title":"First Kate post"}]}}}}]
```

### CODE REVIEW

File: src/PrefetchBuffer.php [hunk 1]
```diff
@@ -20,14 +22,27 @@ class PrefetchBuffer
     private array $results = [];
 
     /** @param array<array-key, mixed> $arguments The input arguments passed from GraphQL to the field. */
-    public function register(object $object, array $arguments): void
-    {
-        $this->objects[$this->computeHash($arguments)][] = $object;
+    public function register(
+        object $object,
+        array $arguments,
+        mixed $info = null,
```
Comment (grynchuk) [inline src/PrefetchBuffer.php:28]: Should be ?ResolveInfo to avoid BC break, but it is disallowed by cs.

**Review by oojacoboo — CHANGES_REQUESTED** [review]
Thanks for the PR @grynchuk!  Everything looks pretty good - some CS fixes needed.

File: tests/Fixtures/Integration/Models/Company.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TheCodingMachine\GraphQLite\Fixtures\Integration\Models;
+
+use TheCodingMachine\GraphQLite\Annotations\Type;
+
+/**
+ * @Type()
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Models/Company.php]: fixed

File: tests/Fixtures/Integration/Types/ContactType.php [hunk 3]
```diff
@@ -53,4 +54,55 @@ public static function prefetchContacts(iterable $contacts, string $prefix)
...
+    public static function prefetchPosts(iterable $contacts): array
+    {
+        $posts = [];
+        foreach ($contacts as $contact) {
+            $contactPost = array_filter(
+                self::getContactPosts(),
+                fn(Post $post) => $post->author?->getName() === $contact->getName()
+            );
+
+            if ([] === $contactPost) {
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Types/ContactType.php]: fixed

File: tests/Fixtures/Integration/Types/PostType.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace TheCodingMachine\GraphQLite\Fixtures\Integration\Types;
+
+use TheCodingMachine\GraphQLite\Annotations\ExtendType;
+use TheCodingMachine\GraphQLite\Annotations\Field;
+use TheCodingMachine\GraphQLite\Fixtures\Integration\Models\Post;
+
+/**
+ * @ExtendType(class=Post::class)
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Types/PostType.php]: fixed

File: tests/Fixtures/Integration/Controllers/CompanyController.php [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+declare(strict_types=1);
+
+namespace TheCodingMachine\GraphQLite\Fixtures\Integration\Controllers;
+
+use TheCodingMachine\GraphQLite\Annotations\Query;
+use TheCodingMachine\GraphQLite\Fixtures\Integration\Models\Company;
+
+class CompanyController
+{
+    /** @Query() */
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Controllers/CompanyController.php]: ok

File: tests/Fixtures/Integration/Types/ContactType.php [hunk 6]
```diff
@@ -53,4 +54,55 @@ public static function prefetchContacts(iterable $contacts, string $prefix)
             'prefix' => $prefix
         ];
     }
+
+    /**
+     * @Field()
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Types/ContactType.php]: fixed

File: tests/Fixtures/Integration/Types/CompanyType.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+namespace TheCodingMachine\GraphQLite\Fixtures\Integration\Types;
+
+use TheCodingMachine\GraphQLite\Annotations\ExtendType;
+use TheCodingMachine\GraphQLite\Annotations\Field;
+use TheCodingMachine\GraphQLite\Annotations\Prefetch;
+use TheCodingMachine\GraphQLite\Fixtures\Integration\Models\Company;
+use TheCodingMachine\GraphQLite\Fixtures\Integration\Models\Contact;
+
+/**
+ * @ExtendType(class=Company::class)
```
[the remark this answers is not in the document]
Comment (grynchuk) [reply tests/Fixtures/Integration/Types/CompanyType.php]: fixed

**Review by oojacoboo — APPROVED** [review]

**Comment by andrew-demb** [comment]
@oojacoboo These fixes merged for the `7.x` version of graphqlite, but there's a major performance problem with `7.x` which is blocked update for `7.x` for the Symfony bundle -  https://github.com/thecodingmachine/graphqlite-bundle/pull/203#issuecomment-2201182492
And issue in graphqlite - https://github.com/thecodingmachine/graphqlite/issues/691

Would you approve PR with the same changes, but for `6.x` branch of graphqlite (and tag an additional `6.x` release after merge)?

**Comment by oojacoboo** [comment]
@andrew-demb maybe it's just a better idea to test the changes in `class-finder` as per the comments in #691 and get that resolved, instead of just back patching, to prevent resolving the actual issue, assuming that you're even seeing any degraded performance.

**Pull request merged**